service-worker/worker/src/plugins/dynamic/index.ts

```diff
@@ -0,0 +1,231 @@
+import {FetchInstruction, NgSwCache, Operation, Plugin, PluginFactory, VersionWorker} from '@angular/service-worker/worker';
+import {Observable} from 'rxjs/Observable';
+
+const DEFAULT_NETWORK_WAIT_TIME_MS = 30000;
+
+export function DynamicContentCache(options?: DynamicContentCacheOptions): PluginFactory<DynamicImpl> {
+  return (worker: VersionWorker) => new DynamicImpl(worker, options.manifestKey || 'dynamic');
+}
+
+export type CachingStrategy = "cacheFirst" | "cacheOnly" | "fastest" | "networkFirst" | "networkOnly";
+
+export interface DynamicContentCacheOptions {
+  manifestKey?: string;
+}
+
+interface DynamicContentCacheManifest {
+  match: UrlMatchConfig[];
+}
+
+interface RegexpMap {
+  [str: string]: RegExp;
+}
+
+export interface UrlMatchOrInvalidateConfig {
+  regex?: string;
+  url?: string;
+  prefix?: boolean;
+}
+
+export interface UrlMatchConfig extends UrlMatchOrInvalidateConfig {
+  strategy: CachingStrategy;
+  waitForNetworkMs?: number;
+  invalidates?: UrlMatchOrInvalidateConfig[];
+}
+
+export class DynamicImpl implements Plugin<DynamicImpl> {
+  errNotFound: Response;
+  errGatewayTimeout: Response;
+  cacheKey: string;
+
+  private regexpMap = {};
+
+  constructor(
+    private worker: VersionWorker,
+    private manifestKey: string) {
+      this.cacheKey = manifestKey === 'dynamic' ? 'dynamic' : `dynamic:${manifestKey}`;
+      this.errGatewayTimeout = 
+          this.worker.adapter.newResponse('Gateway Timeout', 504, 'Gateway Timeout');
+      this.errNotFound =
+        this.worker.adapter.newResponse('Not Found', 404, 'Not Found');
+  }
+
+  private get config(): DynamicContentCacheManifest {
+    return this.worker.manifest[this.manifestKey] as DynamicContentCacheManifest;
+  }
+
+  private regexp(str: string): RegExp {
+    if (!this.regexpMap.hasOwnProperty(str)) {
+      this.regexpMap[str] = new RegExp(str).compile();
```

Comment:
Is the usage of `compile()` intentional? It seems [deprecated](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/compile).